File: scripts/test_classification.py
#!/usr/bin/env python3
"""Replay hook data log against a different prompt/model for prompt tuning.

Reads the JSONL hook data log and re-runs fact extraction with a specified
prompt and model, comparing results to the logged decisions.

Usage:
    uv run scripts/test_classification.py --prompt PATH --model MODEL [--data PATH] [--ollama-url URL]
"""

from __future__ import annotations

import argparse
import json
import sys
import time
from datetime import datetime
from pathlib import Path
from string import Template

import httpx

_DEFAULT_DATA = Path.home() / ".claude" / "logs" / "mindojo-hook-data.jsonl"
_DEFAULT_OLLAMA_URL = "http://localhost:11434"


def _load_env(repo_root: Path) -> dict[str, str]:
    """Parse key=value pairs from .env file (no shell expansion)."""
    env_path = repo_root / ".env"
    if not env_path.is_file():
        return {}
    result: dict[str, str] = {}
    for line in env_path.read_text().splitlines():
        line = line.strip()
        if not line or line.startswith("#"):
            continue
        if "=" not in line:
            continue
        key, _, value = line.partition("=")
        key = key.strip()
        value = value.strip().strip("\"'")
        if key:
            result[key] = value
    return result


def _call_ollama(
    url: str,
    model: str,
    system_prompt: str,
    content: str,
    *,
    api_key: str = "",
    max_attempts: int = 3,
    timeout: float = 120.0,
) -> list[str] | None:
    """Send content to Ollama and parse facts. Returns None on failure."""
    headers: dict[str, str] = {"Content-Type": "application/json"}
    if api_key:
        headers["Authorization"] = f"Bearer {api_key}"

    payload = {
        "model": model,
        "messages": [
            {"role": "system", "content": system_prompt},
            {"role": "user", "content": content},
        ],
        "stream": False,
        "format": "json",
        "options": {"num_predict": 1024},
        "think": False,
    }

    for attempt in range(1, max_attempts + 1):
        try:
            resp = httpx.post(
                f"{url}/api/chat",
                json=payload,
                headers=headers,
                timeout=timeout,
            )
            resp.raise_for_status()
            data = resp.json()
            text = data.get("message", {}).get("content", "")
            parsed = json.loads(text)
            return parsed.get("facts", [])
        except (httpx.TimeoutException, httpx.ConnectError) as exc:
            if attempt < max_attempts:
                print(
                    f"  Attempt {attempt}/{max_attempts} failed ({exc}), retrying in 5s..."
                )
                time.sleep(5)
            else:
                print(f"  All {max_attempts} attempts failed: {exc}")
                return None
        except (json.JSONDecodeError, KeyError) as exc:
            print(f"  Parse error: {exc}")
            return None
        except httpx.HTTPStatusError as exc:
            print(f"  HTTP error: {exc}")
            return None
    return None


def main() -> None:
    parser = argparse.ArgumentParser(
        description="Replay hook data log with a different prompt/model"
    )
    parser.add_argument(
        "--prompt",
        required=True,
        type=Path,
        help="Path to fact-extraction prompt .md file",
    )
    parser.add_argument(
        "--model", required=True, help="Ollama model name (e.g. qwen3.5:4b)"
    )
    parser.add_argument(
        "--data",
        type=Path,
        default=_DEFAULT_DATA,
        help=f"Path to JSONL data file (default: {_DEFAULT_DATA})",
    )
    parser.add_argument(
        "--ollama-url",
        default=None,
        help="Ollama API URL (default: from .env or localhost)",
    )
    parser.add_argument(
        "--ollama-api-key", default=None, help="Ollama API key (default: from .env)"
    )
    args = parser.parse_args()

    repo_root = Path(__file__).resolve().parent.parent
    env_vars = _load_env(repo_root)

    ollama_url = args.ollama_url or env_vars.get("OLLAMA_URL", _DEFAULT_OLLAMA_URL)
    api_key = args.ollama_api_key or env_vars.get("OLLAMA_API_KEY", "")

    if not args.prompt.is_file():
        print(f"Error: prompt file not found: {args.prompt}", file=sys.stderr)
        sys.exit(1)

    if not args.data.is_file():
        print(f"Error: data file not found: {args.data}", file=sys.stderr)
        sys.exit(1)

    prompt_template = args.prompt.read_text()
    today = datetime.now().strftime("%Y-%m-%d")
    system_prompt = Template(prompt_template).safe_substitute(today=today)

    entries = []
    for line in args.data.read_text().splitlines():
        line = line.strip()
        if not line:
            continue
        try:
            entry = json.loads(line)
            decision = entry.get("decision", "")
            if decision in ("kept", "rejected"):
                entries.append(entry)
        except json.JSONDecodeError:
            continue

    if not entries:
        print("No testable entries found (need 'kept' or 'rejected' decisions)")
        sys.exit(0)

    print(f"Loaded {len(entries)} testable entries from {args.data}")
    print(f"Model: {args.model}")
    print(f"Prompt: {args.prompt}")
    print(f"Ollama: {ollama_url}")
    print()

    tp = fp = tn = fn = 0

    for i, entry in enumerate(entries, 1):
        logged_decision = entry["decision"]
        content = entry["content"]
        expect_facts = logged_decision == "kept"

        new_facts = _call_ollama(
            url=ollama_url,
            model=args.model,
            system_prompt=system_prompt,
            content=content,
            api_key=api_key,
        )

        if new_facts is None:
            status = "ERROR"
            label = "---"
        else:
            got_facts = len(new_facts) > 0
            if expect_facts and got_facts:
                tp += 1
                status = "TP"
                label = "KEPT"
            elif expect_facts and not got_facts:
                fn += 1
                status = "FN"
                label = "KEPT"
            elif not expect_facts and not got_facts:
                tn += 1
                status = "TN"
                label = "REJECTED"
            else:
                fp += 1
                status = "FP"
                label = "REJECTED"

        prefix = content[:60].replace("\n", " ")
        n_facts = len(new_facts) if new_facts is not None else "?"
        print(f"[{i}/{len(entries)}] {status} {label} facts={n_facts} | {prefix}...")

    print()
    print("=" * 60)
    total = tp + fp + tn + fn
    accuracy = (tp + tn) / total * 100 if total else 0
    precision = tp / (tp + fp) * 100 if (tp + fp) else 0
    recall = tp / (tp + fn) * 100 if (tp + fn) else 0

    print(f"Total entries:  {total}")
    print(f"True Positive:  {tp}")
    print(f"False Positive: {fp}")
    print(f"True Negative:  {tn}")
    print(f"False Negative: {fn}")
    print(f"Accuracy:       {accuracy:.1f}%")
    print(f"Precision:      {precision:.1f}%")
    print(f"Recall:         {recall:.1f}%")


if __name__ == "__main__":
    main()

File: rs/memcan-server/src/test_classification.rs
//! Replay hook data log for prompt tuning (moved from `memcan-test-classification`).

use chrono::Utc;
use serde::Deserialize;

use memcan_core::config::Settings;
use memcan_core::error::{MemcanError, Result as MemcanResult, ResultExt};
use memcan_core::llm::GenaiLlmProvider;
use memcan_core::traits::{LlmMessage, LlmOptions, LlmProvider, Role};

use crate::TestClassificationArgs;

fn default_data_path() -> std::path::PathBuf {
    dirs::home_dir()
        .unwrap_or_else(|| std::path::PathBuf::from("/tmp"))
        .join(".claude")
        .join("logs")
        .join("memcan-hook-data.jsonl")
}

#[derive(Deserialize)]
struct HookDataEntry {
    #[serde(default)]
    decision: String,
    #[serde(default)]
    content: String,
}

#[derive(Deserialize)]
struct FactsResponse {
    #[serde(default)]
    facts: Vec<String>,
}

async fn call_llm(
    llm: &dyn LlmProvider,
    model: &str,
    system_prompt: &str,
    content: &str,
    max_attempts: u32,
) -> Option<Vec<String>> {
    let messages = vec![
        LlmMessage {
            role: Role::System,
            content: system_prompt.to_string(),
        },
        LlmMessage {
            role: Role::User,
            content: content.to_string(),
        },
    ];

    let options = Some(LlmOptions {
        format_json: true,
        max_tokens: Some(1024),
        ..Default::default()
    });

    for attempt in 1..=max_attempts {
        match llm.chat(model, &messages, options.clone()).await {
            Ok(text) => match serde_json::from_str::<FactsResponse>(&text) {
                Ok(parsed) => return Some(parsed.facts),
                Err(e) => {
                    println!("  Parse error: {}", e);
                    return None;
                }
            },
            Err(e) => {
                if attempt < max_attempts {
                    println!(
                        "  Attempt {}/{} failed ({}), retrying in 5s...",
                        attempt, max_attempts, e
                    );
                    tokio::time::sleep(std::time::Duration::from_secs(5)).await;
                } else {
                    println!("  All {} attempts failed: {}", max_attempts, e);
                    return None;
                }
            }
        }
    }

    None
}

pub async fn run(args: &TestClassificationArgs) -> MemcanResult<()> {
    let settings = Settings::load()?;
    settings.ensure_log_dir()?;
    let llm = GenaiLlmProvider::from_settings(&settings);

    if !args.prompt.is_file() {
        return Err(MemcanError::Other(format!(
            "Prompt file not found: {}",
            args.prompt.display()
        )));
    }

    let data_path = args.data.clone().unwrap_or_else(default_data_path);

    if !data_path.is_file() {
        return Err(MemcanError::Other(format!(
            "Data file not found: {}",
            data_path.display()
        )));
    }

    let prompt_template = std::fs::read_to_string(&args.prompt)
        .with_context(|| format!("failed to read prompt: {}", args.prompt.display()))?;

    let today = Utc::now().format("%Y-%m-%d").to_string();
    let system_prompt = prompt_template.replace("$today", &today);

    let data_raw = std::fs::read_to_string(&data_path)
        .with_context(|| format!("failed to read data: {}", data_path.display()))?;

    let mut entries: Vec<HookDataEntry> = Vec::new();
    for line in data_raw.lines() {
        let line = line.trim();
        if line.is_empty() {
            continue;
        }
        match serde_json::from_str::<HookDataEntry>(line) {
            Ok(entry) => {
                if entry.decision == "kept" || entry.decision == "rejected" {
                    entries.push(entry);
                }
            }
            Err(_) => continue,
        }
    }

    if entries.is_empty() {
        println!("No testable entries found (need 'kept' or 'rejected' decisions)");
        return Ok(());
    }

    println!(
        "Loaded {} testable entries from {}",
        entries.len(),
        data_path.display()
    );
    println!("Model: {}", args.model);
    println!("Prompt: {}", args.prompt.display());
    println!();

    let mut tp: u32 = 0;
    let mut fp: u32 = 0;
    let mut tn: u32 = 0;
    let mut r#fn: u32 = 0;

    for (i, entry) in entries.iter().enumerate() {
        let expect_facts = entry.decision == "kept";
        let new_facts = call_llm(&llm, &args.model, &system_prompt, &entry.content, 3).await;

        let (status, label) = match &new_facts {
            None => ("ERROR", "---"),
            Some(facts) => {
                let got_facts = !facts.is_empty();
                if expect_facts && got_facts {
                    tp += 1;
                    ("TP", "KEPT")
                } else if expect_facts && !got_facts {
                    r#fn += 1;
                    ("FN", "KEPT")
                } else if !expect_facts && !got_facts {
                    tn += 1;
                    ("TN", "REJECTED")
                } else {
                    fp += 1;
                    ("FP", "REJECTED")
                }
            }
        };

        let prefix: String = entry
            .content
            .chars()
            .take(60)
            .collect::<String>()
            .replace('\n', " ");
        let n_facts = new_facts
            .as_ref()
            .map(|f| f.len().to_string())
            .unwrap_or_else(|| "?".to_string());

        println!(
            "[{}/{}] {} {} facts={} | {}...",
            i + 1,
            entries.len(),
            status,
            label,
            n_facts,
            prefix
        );
    }

    println!();
    println!("{}", "=".repeat(60));

    let total = tp + fp + tn + r#fn;
    let accuracy = if total > 0 {
        (tp + tn) as f64 / total as f64 * 100.0
    } else {
        0.0
    };
    let precision = if (tp + fp) > 0 {
        tp as f64 / (tp + fp) as f64 * 100.0
    } else {
        0.0
    };
    let recall = if (tp + r#fn) > 0 {
        tp as f64 / (tp + r#fn) as f64 * 100.0
    } else {
        0.0
    };

    println!("Total entries:  {}", total);
    println!("True Positive:  {}", tp);
    println!("False Positive: {}", fp);
    println!("True Negative:  {}", tn);
    println!("False Negative: {}", r#fn);
    println!("Accuracy:       {:.1}%", accuracy);
    println!("Precision:      {:.1}%", precision);
    println!("Recall:         {:.1}%", recall);

    Ok(())
}
